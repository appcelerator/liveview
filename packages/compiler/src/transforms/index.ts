export { I18nTransform } from './i18n';
export { TranspileTransform } from './transpile';
export { Transform } from './transform';
