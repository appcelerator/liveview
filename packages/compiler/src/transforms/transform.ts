export interface Transform {
  apply(file: string, content: string): Promise<string>
}
