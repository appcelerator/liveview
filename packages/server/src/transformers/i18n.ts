import xmlParser from 'fast-xml-parser';
import fs from 'fs-extra';
import path from 'path';

import { TransferInfo } from "../index";
import { AbstractTransformer, TransformerOptions } from "./base";

interface Translations {
  [key: string]: string
}

interface I18nDict {
  [key: string]: Translations
}

const localePattern = /i18n[\\/](.+)[\\/]strings\.xml$/;

/**
 * Transformer for the localization files under the project's `i18n` directory.
 */
export class I18nTransformer extends AbstractTransformer {
  private lang: I18nDict = {}

  constructor(options: TransformerOptions) {
    super(options);
  }

  async transform(changes: string[], removals?: string[]): Promise<TransferInfo[]> {
    const files: TransferInfo[] = [];
    if (removals) {
      for (const file of removals) {
        const langCode = this.getLanguageCode(file);
        if (langCode) {
          delete this.lang[langCode];
        }
      }
    }

    for (const file of changes) {
      const translations: Translations = {};
      const langCode = this.getLanguageCode(file);
      if (langCode) {
        const content  = await fs.readFile(file, 'utf-8');
        const result = xmlParser.parse(content, {
          ignoreAttributes: false
        });
        if (!result.resources) {
          continue;
        }
        let stringNodes = result.resources.string || [];
        if (!Array.isArray(stringNodes)) {
          stringNodes = [stringNodes];
        }
        for (const node of stringNodes) {
          const key = node['@_name'];
          const value = node['#text'];
          translations[key] = value;
        }
        this.lang[langCode] = translations;
      }
    }

    const outputFile = path.join(this.intermediateBuildDir, '__i18n__.json');
    await fs.writeJson(outputFile, this.lang);

    files.push({
      from: path.relative(this.workspacePath, outputFile),
      to: '__i18n__.json'
    });

    return files;
  }

  private getLanguageCode(file: string): string | null {
    const match = file.match(localePattern);
    if (!match) {
      return null;
    }

    return match[1];
  }
}
