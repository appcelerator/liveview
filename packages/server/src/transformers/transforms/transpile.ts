import path from 'path';
import jsanalyze from 'node-titanium-sdk/lib/jsanalyze';

import { AssetTransform } from '../base';

interface TranspileTargets {
  [key: string]: string
}

export default class TranspileTransform implements AssetTransform {
  constructor(private targets: TranspileTargets) {

  }

  apply(file: string): string {
    const filename = path.basename(file);
    const transpiled = jsanalyze.analyzeJsFile(file, {
      filename: filename,
      minify: false,
      transpile: true,
      targets: this.targets,
      sourceMap: false
    });
    return transpiled.contents;
  }
}
