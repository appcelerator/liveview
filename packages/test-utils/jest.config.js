const path = require('path');

module.exports = {
  preset: 'ts-jest',
  rootDir: '../../',
  testEnvironment: 'node',
  testMatch: [path.join(__dirname, 'test', '**/?(*.)spec.ts')],
  moduleNameMapper: {
    '^@liveview/(.*)$': '<rootDir>/packages/$1/src'
  }
};
