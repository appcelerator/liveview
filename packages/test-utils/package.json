{
  "name": "@liveview/test-utils",
  "version": "1.0.0",
  "main": "./dist/index.js",
  "scripts": {
    "build": "tsc -b tsconfig.build.json",
    "dev": "tsc -b tsconfig.build.json --watch"
  },
  "author": "Axway Appcelerator",
  "maintainers": [
    "Jan Vennemann <jvennemann@axway.com>"
  ],
  "license": "Apache-2.0",
  "files": [
    "dist",
    "fixtures"
  ],
  "dependencies": {
    "fs-extra": "^9.0.1",
    "tempy": "^1.0.0"
  }
}
