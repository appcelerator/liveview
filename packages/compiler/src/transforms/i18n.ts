import xmlParser from 'fast-xml-parser';
import { Transform } from './transform';

interface Translations {
  [key: string]: string
}

/**
 * Transform for the localization files under the project's `i18n` directory.
 */
export class I18nTransform implements Transform {
  async apply(file: string, content: string): Promise<string> {
    const translations: Translations = {};

    const result = xmlParser.parse(content, {
      ignoreAttributes: false
    });
    if (!result.resources) {
      return JSON.stringify(translations);
    }
    let stringNodes = result.resources.string || [];
    if (!Array.isArray(stringNodes)) {
      stringNodes = [stringNodes];
    }
    for (const node of stringNodes) {
      const key = node['@_name'];
      const value = node['#text'];
      translations[key] = value;
    }

    return JSON.stringify(translations);
  }
}
