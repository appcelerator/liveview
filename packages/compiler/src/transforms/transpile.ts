import { TranspileTargets } from '@liveview/shared-utils';
import jsanalyze = require('node-titanium-sdk/lib/jsanalyze');

import { Transform } from "./transform";

export interface TranspileTransformOptions {
  targets: TranspileTargets
}

let _jsAnalyze: typeof jsanalyze;

/**
 * Transpiles .js files using jsanalyze.
 */
export class TranspileTransform implements Transform {
  private targets: TranspileTargets

  constructor(options: TranspileTransformOptions) {
    this.targets = options.targets;

    _jsAnalyze = require('node-titanium-sdk/lib/jsanalyze');
  }

  async apply(file: string, content: string): Promise<string> {
    const transpiled = _jsAnalyze.analyzeJs(content, {
      filename: file,
      minify: false,
      transpile: true,
      targets: this.targets,
      sourceMap: false
    });
    // transpiled.contents can be empty if the file was ignored via babel config
    return transpiled.contents ? transpiled.contents : content;
  }
}
