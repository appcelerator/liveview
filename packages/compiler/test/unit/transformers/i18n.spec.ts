import { getTemplatePath } from '@liveview/test-utils';
import fs from 'fs-extra';
import path from 'path';

import { I18nTransform } from '@/transforms/i18n';


test('turn xml files into json', async () => {
  const i18n = new I18nTransform();
  const file = path.join(getTemplatePath('classic'), 'i18n', 'en', 'strings.xml');
  const content = await fs.readFile(file, 'utf-8');
  const transformed = await i18n.apply(file, content);
  expect(transformed).toMatchSnapshot();
});
