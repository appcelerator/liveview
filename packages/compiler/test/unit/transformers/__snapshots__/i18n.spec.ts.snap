// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`turn xml files into json 1`] = `"{\\"greeting\\":\\"Howdy\\"}"`;
